Add recursive usage analysis

Gbury opened the pull request:
This PR adds a recursive usage analysis. It is split in a few parts:
- on the way down, we record information relative to usage of variables: let-bindings, continuation bindings (to get the parameter list), continuation applications. This enable to build a graph of variables to variables, such that there is an edge [src -> dst], iff usage of `src` implies usage of `dst`.
- once we're at the very bottom of the traversal, we perform a reachability analysis on the graph built, starting from a set of originally used variables which contains arguments passed to the return continuation of the function, arguments of primitive that have effects, and arguments of function calls
- The results of that reachability analysis (i.e. the set of all variables actually used in the whole function) are then stored in the upwards_acc as `used_continuation_params` (which may be a bit of a misnomer actually)
- We can then use that information to filter out the params (and extra_params) of continuations
- Finally, there is a subtlety in the case of recursive continuations: since we can modify the actual parameters used by a continuation, we have to generate a corresponding `apply_cont_rewrite` when we want to change the parameters. This is easy with non-recursive continuation since once we have finished the upwards traversal of the continuation's handler, we can generate the rewrite and add it for the traversal of the continuation-binding body. However in the cas of recursive continuations, we need the rewrite to also apply to the travsersal of the continuation's handler, since there is very likely some recursive calls that need to be rewritten. That means that we need to create the rewrite rule before going up the continuation handler. This is the reason why we need the recursive usage analysis performed globally between the down an up traversals.

chambart commented on middle_end/flambda/simplify/env/downwards_acc.ml:
```diff
@@ -184,3 +196,5 @@ let all_continuations_used t =
 let with_used_closure_vars t ~used_closure_vars =
   { t with used_closure_vars = used_closure_vars; }
+
+
```
You lefts some spurious white space changes here and on a few other files

Gbury replied:
I removed the spurious whitespace diffs

chambart commented on middle_end/flambda/simplify/env/downwards_acc.mli:
```diff
@@ -40,6 +40,12 @@ val map_denv
+(** Extract the rec usage accumulator *)
+val rec_uses : t -> Rec_uses.t
+
+(** Map the rec_uses of the gicen downwards accumulator. *)
```
gicen => given

Gbury replied:
fixed.

chambart commented on middle_end/flambda/simplify/env/rec_uses.mli:
```diff
@@ -0,0 +1,83 @@
+  exn_continuation:Continuation.t ->
+  t -> Variable.Set.t
+(** Analyze the rec uses.
+    TODO: have a more useful return type. *)
```
Either rename the function to something like `live_variables` or change the return type to
```OCaml
type result = {
  live_variables : Variable.Set.t;
}
```

Gbury replied:
I changed the return type.

chambart commented on middle_end/flambda/simplify/env/rec_uses.mli:
```diff
@@ -0,0 +1,83 @@
+(** Pop the current top of the stack. Used when exiting the current
+    contionuation handler. *)
+
+val add_binding : Variable.t -> Name_occurrences.t -> t -> t
```
`record_...` would probably be better names than `add_...` for those functions. I'm not certain.

Gbury replied:
Indeed `add_` might not be the best name, @lthls any ideas/suggestions ?

lthls replied:
`enter_binding` ?

Gbury replied:
it's not really correct to say `enter` here I'd say (as oppposed to say `enter_continuation`); as Pierre said, it's more of a "record that this binding is here at this point"

chambart commented on middle_end/flambda/simplify/env/rec_uses.mli:
```diff
@@ -0,0 +1,83 @@
+(** Initialize the rec_uses so that the stack consists of a single
+    toplevel continuation. *)
+
+val stack_cont : Continuation.t -> Variable.t list -> t -> t
```
The implementation shouldn't matter, it should maybe be called `enter_continuation_handler` / `leave_continuation_handler`

Gbury replied:
I have renamed these functions (though using `exit` rather than `leave`, i'm not sure if the difference matters).

chambart commented on middle_end/flambda/simplify/env/rec_uses.mli:
```diff
@@ -0,0 +1,83 @@
+(*                                                                        *)
+(**************************************************************************)
+
+(** Rec uses
```
A posteriori, I'm not convinced by the name of this module. It is used to remove unused arguments of recursive continuations. But the uses themselves are not really recursive. Well, I don't really have a good name to propose either.
`Variable_dependency_graph` (maybe not, we might end up using that for symbols also), `Uses` ?

Gbury replied:
`Uses` would work, @lthls @mshinwell do you agree ?

lthls replied:
I don't like `Uses`, it's too generic. `Name_uses` would be ok, even though it tracks more than `Name.t`s, it's coherent with the various modules in `naming`.

mshinwell replied:
I agree that `Uses` is too generic.  However let's make sure the name reflects accurately what is going on, so probably not `Name_uses` either.

Gbury replied:
Interrestingly, the information recorded in the module is enough to later compute a dominator (and even pdce I'd say) analysis, since we have the dependency graphs of all variables, as well as which variables are used and bound in which continuation, and which continuation calls which other one. I don't know if that helps with finding a good module name though.

Gbury replied:
@chambart  suggests `Data_flow_graph`, or `Data_flow`, based on the explanation above.

chambart commented on middle_end/flambda/simplify/env/rec_uses.ml:
```diff
@@ -0,0 +1,337 @@
+     CR gbury: would using sets (and thus union and diff) instead of a
+               queue be better ? Enqueuing new nodes would be faster, but
+               we'd lose the bread-first aspect, and revert to some kind of
+               lexicographic order of traversal. *)
```
I don't think the traversal order matters, but I also don't think that using sets instead of a queue for the pending nodes would be faster. But maybe changing
```OCaml
        Variable.Set.iter (fun dst ->
          if not (Variable.Set.mem dst visited) then Queue.push dst queue
        ) (edges t ~src:v);
```
into
```OCaml
        Variable.Set.iter (fun dst -> Queue.push dst queue)
          (Variable.Set.diff (edges t ~src:v) visited);
```
Could be faster by benefiting the of the efficient patricia tree implementation of diff.
Also we might want to avoid checking twice for the presence in the visited set, by considering only an 'enqueued' set.

Gbury replied:
The code now use the set diff. I'm not sure what you mean by an `enqueued` set, or rather what the difference with turning the `queue` into a set would be.

chambart commented on middle_end/flambda/simplify/env/rec_uses.ml:
```diff
@@ -0,0 +1,337 @@
+        ) graph extra_params_and_args.extra_params extra_args
+      ) extra_params_and_args.extra_args graph
+    ) extra graph
+  in
```
Maybe we could extract the graph building part into its own function to allow exposing the intermediate type to allow printing it for debug. This would add to the module interface
```OCaml
type dependency_graph = {
  dependency : Variable.Set.t Variable.Map.t;
  unconditionnaly_used : Variable.Set.t;
}
val build_dependency_graph : t -> dependency_graph
val print_graph : ppf -> dependency_graph -> unit
val required_variables : dependency_graph -> Variable.Set.t
```

Gbury replied:
I split the code into a `Dependency_graph` submodule with your suggested type, the code is indeed slightly easier to read and understand now.

mshinwell replied:
Also: unconditionnaly_used -> unconditionally_used

Gbury replied:
fixed.

chambart commented on middle_end/flambda/simplify/env/upwards_acc.ml:
```diff
@@ -32,20 +32,22 @@ type t = {
   name_occurrences : Name_occurrences.t;
   used_closure_vars : Name_occurrences.t;
   shareable_constants : Symbol.t Static_const.Map.t;
+  used_continuation_params : Variable.Set.t;
```
What is contained here is actually more that just the continuations parameters, it's all the variables that are marked as required by rec_uses. But we only use it for continuation parameters... So I don't know if it need any change.

Gbury replied:
I renamed it to `required_variables` as it seems more appropriate

chambart commented on middle_end/flambda/simplify/simplify_toplevel.rec.ml:
```diff
@@ -20,16 +20,24 @@ open! Simplify_import
 
 let simplify_toplevel dacc expr ~return_continuation ~return_arity
       exn_continuation ~return_cont_scope ~exn_cont_scope =
+  let toplevel_cont = Continuation.create ~name:"toplevel" () in
```
Maybe make it explicit that this is not a real thing by calling it `dummy_toplevel_continuation` and add a comment that this is only required by `Rec_uses`

Gbury replied:
done

chambart reviewed (commented):
Right now functions force the return continuation argument to be kept alive to prevent removing it which would produce mismatch continuation arity. But this prevents some dead code to be eliminated
```OCaml
let f g b x =
  let x =
    if b then g x
    else x+1
  in
  ignore x
```
We would like `x+1` to be replaced by a dummy expression instead of just kept.

I don't suggest that we do this change right now, but we might want to add a CR and example of a problematic example to flambdatest at least.

lpw25 commented:
I don't quite understand your return continuation example. Surely `ignore` as a primitive gets replaced by a fresh `()` unrelated to `x`? So what is keeping `x` alive?

Gbury commented:
As @lpw25 said, the example suggested by @chambart  compiles to:
```
(λ〈k33〉《k34》(g/52 ∷ 𝕍) (b/53 ∷ 𝕍) (x/54 ∷ 𝕍=tagged_ℕ𝕚) (my_closure/55 ∷ 𝕍) .
    ((let
      (untagged/56 = ((Untag_imm b/53)))
      (switch untagged/56
      | 0 ↦ goto k35
      | 1 ↦ goto k36))
     k36:
      ((apply
        ((g/52〈k37〉《k34》 (x/54)) (call_kind Indirect_unknown_arity)
         (dbg /home/guigui/tmp/unused.ml:3,14--17) (inline Default_inline)
         (inlining_state (depth 0))))
       k37 (param/57 ∷ 𝕍): goto k35)
     k35: return k33 0))))
```

Which actually did exactly what one could expect (remove the unused `x + 1`, and use an intermediary continuation to "ignore" the result of the function application). That said, maybe there are more complex examples that exhibit the problematic behaviour ?

lthls commented on middle_end/flambda/simplify/env/rec_uses.mli:
```diff
@@ -0,0 +1,88 @@
+
+val exit_continuation : Continuation.t -> t -> t
+(** Pop the current top of the stack. Used when exiting the current
+    contionuation handler. *)
```
```suggestion
    continuation handler. *)
```

lthls commented on middle_end/flambda/simplify/env/rec_uses.ml:
```diff
@@ -0,0 +1,375 @@
+  (* breadth-first reachability analysis
+     CR gbury: would using sets (and thus union and diff) instead of a
+               queue be better ? Enqueuing new nodes would be faster, but
+               we'd lose the bread-first aspect, and revert to some kind of
```
Always go for the bread first !

Gbury replied:
I've switched to using an `enqueued` set rather than the `visited` one, and removed the comment, ^^

Gbury commented:
I renamed the `Rec_uses` module to `Data_flow`, which is more appropriate, and rebased the branch, so this can be merged as soon as CI is green and people have had a last look.

chambart reviewed (approved):
I think this is now good.

mshinwell commented:
I improved the efficiency of the code in `Simplify_let`, threaded through the `generate_phantom_lets` setting from `DE` instead of recomputing it, and made some other various tidyups.  OK to merge when CI passes.
